In [96]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
data = pd.read_csv(r"C:\Users\User\Downloads\TITANIC.csv")
data

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [97]:
#Finding missing data
data.isnull()

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket  \
0          False     False   False  False  False  False  False  False   False   
1          False     False   False  False  False  False  False  False   False   
2          False     False   False  False  False  False  False  False   False   
3          False     False   False  False  False  False  False  False   False   
4          False     False   False  False  False  False  False  False   False   
..           ...       ...     ...    ...    ...    ...    ...    ...     ...   
886        False     False   False  False  False  False  False  False   False   
887        False     False   False  False  False  False  False  False   False   
888        False     False   False  False  False   True  False  False   False   
889        False     False   False  False  False  False  False  False   False   
890        False     False   False  False  False  False  False  False   False   

      Fare  Cabin  Embarked  
0    False   True     False  
1    False  False     False  
2    False   True     False  
3    False  False     False  
4    False   True     False  
..     ...    ...       ...  
886  False   True     False  
887  False  False     False  
888  False   True     False  
889  False  False     False  
890  False   True     False  

[891 rows x 12 columns]

Clearly, some fields have missing data.

In [98]:
data.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [99]:
#creating function to replace missing ages by median of ages
def replace_age(c):
    Age = c[0]
    Pclass = c[1]
    if pd.isnull(Age):
        return int(data[data["Pclass"] == Pclass]["Age"].median())
    else:
        return Age

In [100]:
data["Age"] = data[["Age", "Pclass"]].apply(replace_age, axis=1)

In [101]:
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [102]:
data.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [103]:
data = data.drop(["Cabin"],axis=1)

In [104]:
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked  
0      0         A/5 21171   7.2500        S  
1      0          PC 17599  71.2833        C  
2      0  STON/O2. 3101282   7.9250        S  
3      0            113803  53.1000        S  
4      0            373450   8.0500        S

In [105]:
#impute embark with the most common value found by visual analysis
most_common_value= 'S'

for d in data:
    data['Embarked'] = data['Embarked'].fillna(most_common_value)

In [106]:
data.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64

Data is now cleaned.

In [107]:
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked  
0      0         A/5 21171   7.2500        S  
1      0          PC 17599  71.2833        C  
2      0  STON/O2. 3101282   7.9250        S  
3      0            113803  53.1000        S  
4      0            373450   8.0500        S

In [108]:
data.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Embarked        object
dtype: object

Type conversions need to be performed 

In [109]:
dataTrain = data
dataTest = data

In [110]:
dataTrain.replace(to_replace="male", value=1, inplace=True)
dataTrain.replace(to_replace="female", value=0, inplace=True)

dataTest.replace(to_replace="male", value=1, inplace=True)
dataTest.replace(to_replace="female", value=0, inplace=True)

In [111]:
dataTrain.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name  Sex   Age  SibSp  Parch  \
0                            Braund, Mr. Owen Harris    1  22.0      1      0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...    0  38.0      1      0   
2                             Heikkinen, Miss. Laina    0  26.0      0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)    0  35.0      1      0   
4                           Allen, Mr. William Henry    1  35.0      0      0   

             Ticket     Fare Embarked  
0         A/5 21171   7.2500        S  
1          PC 17599  71.2833        C  
2  STON/O2. 3101282   7.9250        S  
3            113803  53.1000        S  
4            373450   8.0500        S

In [83]:
dataTrain.Age = dataTrain.Age.astype('int')
dataTrain.Fare = dataTrain.Fare.astype('int')
dataTrain.SibSp = dataTrain.SibSp.astype('int')
dataTrain.Parch = dataTrain.Parch.astype('int')
dataTrain.Pclass = dataTrain.Pclass.astype('int')
dataTrain.Sex = dataTrain.Sex.astype('int')
dataTrain.Survived = dataTrain.Survived.astype('int')



In [86]:
#Dropping Name, Ticket, Embarked
dataTrain.drop(["Name"], ["Ticket"], ["Embarked"], axis=1)
dataTest.drop(["Name"], ["Ticket"], ["Embarked"], axis=1)
dataTrain

PassengerId  Survived  Pclass  Sex  Age  SibSp  Parch  Fare
0              1         0       3    1   22      1      0     7
1              2         1       1    0   38      1      0    71
2              3         1       3    0   26      0      0     7
3              4         1       1    0   35      1      0    53
4              5         0       3    1   35      0      0     8
..           ...       ...     ...  ...  ...    ...    ...   ...
886          887         0       2    1   27      0      0    13
887          888         1       1    0   19      0      0    30
888          889         0       3    0   24      1      2    23
889          890         1       1    1   26      0      0    30
890          891         0       3    1   32      0      0     7

[891 rows x 8 columns]

In [87]:
dataTest

PassengerId  Survived  Pclass  Sex  Age  SibSp  Parch  Fare
0              1         0       3    1   22      1      0     7
1              2         1       1    0   38      1      0    71
2              3         1       3    0   26      0      0     7
3              4         1       1    0   35      1      0    53
4              5         0       3    1   35      0      0     8
..           ...       ...     ...  ...  ...    ...    ...   ...
886          887         0       2    1   27      0      0    13
887          888         1       1    0   19      0      0    30
888          889         0       3    0   24      1      2    23
889          890         1       1    1   26      0      0    30
890          891         0       3    1   32      0      0     7

[891 rows x 8 columns]

In [88]:
dataTrain.dtypes

PassengerId    int64
Survived       int32
Pclass         int32
Sex            int32
Age            int32
SibSp          int32
Parch          int32
Fare           int32
dtype: object

Data has been converted to numeric.

Splitting into X ans Y, training and testing data

In [89]:
yTrain = data.pop('Survived').to_numpy()
xTrain = data.to_numpy()
xTest = data.to_numpy()
for i,x in enumerate(yTrain):
    yTrain[i] = float(x)
xTrain.shape, yTrain.shape, xTest.shape

((891, 7), (891,), (891, 7))

In [90]:
xTrain


array([[  1,   3,   1, ...,   1,   0,   7],
       [  2,   1,   0, ...,   1,   0,  71],
       [  3,   3,   0, ...,   0,   0,   7],
       ...,
       [889,   3,   0, ...,   1,   2,  23],
       [890,   1,   1, ...,   0,   0,  30],
       [891,   3,   1, ...,   0,   0,   7]], dtype=int64)

Creating the Model for Logistic Regression

In [115]:
class LogisticRegression:
    
    def __init__(self, learningRate = 0.0001, iterations = 10000):
        #Initialisation     
        self.m = None
        self.n = None
        self.weights = None
        self.bias = None
        self.iterations = iterations
        self.learningRate = learningRate
        
    def lossFunction(self, y, yHat):
       #Binary cross entropy function       
        yZeroLoss = y * np.log(yHat + 1e-9)
        yOneLoss = (1-y) * np.log(1 - yHat + 1e-9)
        return -np.mean(yZeroLoss + yOneLoss)
    
    def sigmoid(self, x):
        #For Probability calculation
        return 1.0/(1+np.exp(-x))
    
    def predict(self, X):
        #If Prob > Threshold, then Label is 1. Otherwise 0.
        return np.array([self.sigmoid(np.dot(x, self.weights)+self.bias) for x in X])
    
    def gradientDescent(self, X, y):       
        for iteration in range(self.iterations):
            y = y.astype(np.float64)
            yHat = self.predict(X)
            
            dw = np.matmul(X.transpose(), yHat-y)
            dw = np.array([np.mean(gradient) for gradient in dw])
            db = np.mean(yHat-y)
            
            #Updation of weights
            
            self.weights -= self.learningRate * dw
            self.bias -= self.learningRate * db
            
            print(f'Iteration: {iteration+1}, Loss: {self.lossFunction(yHat, y)}')
    
    def fit(self, X_train, y_train):        
        self.m , self.n = X_train.shape
        self.weights = np.ones(self.n, dtype=np.float64)
        self.bias = 0
        
        # Gradient descent
        self.gradientDescent(X_train, y_train)
        self.evaluate(X_train, y_train)
            

    
    def predictLabels(self, X, threshold=0.5):
        preds = self.predict(X)
        return np.array([1 if value >= threshold else 0 for value in preds])
    
    def evaluate(self, X_test, y_test, threshold=0.5):
        preds = self.predictLabels(X_test)
        N = y_test.shape[-1]
        accuracy = (y_test == preds).sum() / N
        TP = ((preds == 1) & (y_test == 1)).sum()                     # True positives
        FP = ((preds == 1) & (y_test == 0)).sum()                     # False positives
        precision = TP / (TP+FP)
        
        print(f'Accuracy: {accuracy}, Precision: {precision}')

Fitting the Data into the model

In [116]:
lrClassifier = LogisticRegression(learningRate=0.1, iterations=15)
lrClassifier.fit(xTrain, yTrain)

Iteration: 1, Loss: 12.768880969855868
Iteration: 2, Loss: 7.954384866090542
Iteration: 3, Loss: 7.954384866090542
Iteration: 4, Loss: 12.768880969855868
Iteration: 5, Loss: 7.954384866090542
Iteration: 6, Loss: 7.954384866090542
Iteration: 7, Loss: 12.768880969855868
Iteration: 8, Loss: 7.931126430806652
Iteration: 9, Loss: 12.768880969855868
Iteration: 10, Loss: 7.931126430806652
Iteration: 11, Loss: 8.000901736658323
Iteration: 12, Loss: 12.768880969855868
Iteration: 13, Loss: 7.931126430806652
Iteration: 14, Loss: 7.6520257288139595
Iteration: 15, Loss: 12.768880969855868
Accuracy: 0.6184062850729517, Precision: 0.625


C:\Users\User\AppData\Local\Temp\ipykernel_26596\2306892138.py:20: RuntimeWarning: overflow encountered in exp
  return 1.0/(1+np.exp(-x))


Model works will an accuracy of 61.84% and precision of 62.5%